tools/alive-tv.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2018-present The Alive2 Authors.
+// Distributed under the MIT license that can be found in the LICENSE file.
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Bitcode/BitcodeReader.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/PrettyStackTrace.h"
+#include "llvm/Support/Signals.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/ToolOutputFile.h"
+
+#include "ir/function.h"
+#include "llvm_util/utils.h"
+#include "smt/smt.h"
+#include "smt/solver.h"
+#include "tools/alive_parser.h"
+#include "util/config.h"
+#include "util/file.h"
+#include <cstdlib>
+#include <iostream>
+#include <optional>
+#include <string_view>
+#include <vector>
+
+using namespace IR;
+using namespace tools;
+using namespace util;
+using namespace std;
+using namespace llvm_util;
+
+static llvm::cl::opt<std::string>
+opt_file1(llvm::cl::Positional, llvm::cl::desc("first_bitcode_file"),
+    llvm::cl::Required, llvm::cl::value_desc("filename"));
+
+static llvm::cl::opt<std::string>
+opt_file2(llvm::cl::Positional, llvm::cl::desc("second_bitcode_file"),
+    llvm::cl::Required, llvm::cl::value_desc("filename"));
+
+static llvm::cl::opt<bool> opt_disable_undef("disable-undef-input",
+    llvm::cl::init(false),
+    llvm::cl::desc("Alive: Assume inputs are not undef (default=false)"));
+
+static llvm::cl::opt<bool> opt_disable_poison("disable-poison-input",
+    llvm::cl::init(false),
+    llvm::cl::desc("Alive: Assume inputs are not poison (default=false)"));
+
+static llvm::cl::opt<unsigned> opt_smt_to(
+  "tv-smt-to", llvm::cl::desc("Alive: timeout for SMT queries (default=1000)"),
+  llvm::cl::init(1000), llvm::cl::value_desc("ms"));
+
+static llvm::ExitOnError ExitOnErr;
+
+// adapted from llvm-dis.cpp
+static std::unique_ptr<llvm::Module> openInputFile(llvm::LLVMContext &Context,
+                                                   std::string InputFilename) {
+  std::unique_ptr<llvm::MemoryBuffer> MB =
+    ExitOnErr(errorOrToExpected(llvm::MemoryBuffer::getFile(InputFilename)));
+  std::unique_ptr<llvm::Module> M =
+    ExitOnErr(getOwningLazyBitcodeModule(std::move(MB), Context,
+                                         /*ShouldLazyLoadMetadata=*/true));
+  ExitOnErr(M->materializeAll());
+  return M;
+}
+
+static llvm::Function *getSingleFunction(llvm::Module *M) {
+  llvm::Function *Ret = 0;
+  for (auto &F : *M) {
```

Comment:
This only looks at the actual function definition, it ignores the declarations?
I.e. this should be just one function:
https://godbolt.org/z/RhmJqc

Reply:
sorry, I don't understand what you're getting at.
I only want to look at the function definition, and I only want to look at modules containing a single such definition.

Reply:
That was the question :)
I was just checking that understood correctly that it won't error-out on https://godbolt.org/z/RhmJqc

Reply:
it errors out :)

Reply:
Sounds like a bug then :)

Reply:
fixed thanks